### PR TITLE
Glib: Make libelf dependency optional

### PR DESCRIPTION
**Pull request by Erlkoenig90**
Specify library name and version:  **glib/2.66.0**

This PR makes the glib dependency to libelf truly optional by patching the meson build files ([upstream MR](https://gitlab.gnome.org/GNOME/glib/-/merge_requests/1650)). This way, libelf is not used even if it's available in the system but `with_elf=False` is given.

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
It looks good, let's wait for your EAP approval. Thank you for your contribution!

File: recipes/glib/all/conandata.yml
Comment (ericLemanissier) [inline recipes/glib/all/conandata.yml:18]:
```suggestion
  "2.65.0": []
  "2.65.1": []
  "2.65.2": []
  "2.65.3": []
  "2.66.0":
```

**Review by ericLemanissier — CHANGES_REQUESTED** [review]
you should either make your patch available to other versions of glib, or remove the `with_elf` option if the version is < 2.66.0

File: recipes/glib/all/conandata.yml
Comment (Erlkoenig90) [reply recipes/glib/all/conandata.yml:27]: I made the patch available since 2.65.1 because it works from there. On 2.65.0 we keep the old behaviour.

**Review by ericLemanissier — CHANGES_REQUESTED** [review]
You should del `self.options.with_libelf` in configure if `self.version < tools.Version("2.65.1")`
All other occurrences of `self.options.with_elf` should be replaced with `self.options.get_safe("with_elf", default=True)`

**Comment by Erlkoenig90** [comment]
But then packages that use glib 2.65.0 with `with_elf` set (whatever value) will fail, as that option won't exist anymore.

File: recipes/glib/all/conanfile.py
Comment (SpaceIm) [inline recipes/glib/all/conanfile.py:98]:
```suggestion
        if tools.Version(self.version) >= "2.65.1":
```

File: recipes/glib/all/conanfile.py
Comment (SpaceIm) [inline recipes/glib/all/conanfile.py:108]:
```suggestion
        for patch in self.conan_data.get("patches", {}).get(self.version, []):
            tools.patch(**patch)
```

**Comment by SpaceIm** [comment]
@Erlkoenig90 Please consider https://github.com/Erlkoenig90/conan-center-index/pull/1

**Comment by Erlkoenig90** [comment]
> @Erlkoenig90 Please consider [Erlkoenig90#1](https://github.com/Erlkoenig90/conan-center-index/pull/1)

IMO it looks good, but that will break existing packages that use glib 2.65.0 with `with_elf` set explicitly. Fine by me, but are we okay with that?

**Comment by SpaceIm** [comment]
Oh yes correct, therefore we can delete this option in `configure` method instead of `config_options`, to avoid breaking existing recipe which set `with_elf` explicitly.

**Comment by ericLemanissier** [comment]
IIUK, with_libelf option does not work without your patch, do I'd say a loud break is better than a silent break.

**Comment by Erlkoenig90** [comment]
> IIUC, with_elf option does not work without your patch, do I'd say a loud break is better than a silent break.

It "half-works" as before, by enabling/disabling the conan dependency, and relying on meson to automatically find/not find the libelf dependency when building.

**Comment by SpaceIm** [comment]
I've updated Erlkoenig90#1 to delete `with_elf` in configure

**Comment by Croydon** [comment]
The patch is now part of glib >= 2.67.0

If there is still interest in getting this patched for older versions, the PR needs to resolve the conflicts.

**Comment by ericLemanissier** [comment]
this PR has to be merged anyway, because we need this line:
https://github.com/conan-io/conan-center-index/pull/2925/files#diff-fdf18cd9e1a6791c362f442356c35a16d9ae296ede7454329f7766d60f182fd6R106

**Comment by Erlkoenig90** [comment]
Since the patch is upstream since 2.67.0, I simplified the PR to pass `libelf` to meson when appropriate and removed the patch from the recipe.

File: recipes/glib/all/conanfile.py
Comment (ericLemanissier) [inline recipes/glib/all/conanfile.py:292]:
```suggestion
            )  # this is actually an executable
```
let's correct this typo too ;)

**Review by ericLemanissier — APPROVED** [review]

**Review by prince-chrismc — DISMISSED** [review]

File: recipes/glib/all/conanfile.py
Comment (uilianries) [inline recipes/glib/all/conanfile.py:54]:
```suggestion
        if tools.Version(self.version) < "2.67.0" and not self.options.with_elf:
            raise ConanInvalidConfiguration("libelf dependency can't be disabled in glib < 2.67.0")        
```

This cause should treated on `config_options` instead. As it's an option which is already in the official package, it will better blocking the package id instead of removing it.

File: recipes/glib/all/conanfile.py
Comment (uilianries) [inline recipes/glib/all/conanfile.py:73]: So you can keep the old version

File: recipes/glib/all/conanfile.py
Comment (Erlkoenig90) [reply recipes/glib/all/conanfile.py:54]: Like this?
```python
    def config_options(self):
        [...]
        if tools.Version(self.version) < "2.67.0":
            del self.options.with_elf
```

File: recipes/glib/all/conanfile.py
Comment (Erlkoenig90) [reply recipes/glib/all/conanfile.py:73]: What do you mean?

File: recipes/glib/all/conanfile.py
Comment (uilianries) [reply recipes/glib/all/conanfile.py:54]: Yes, but as this option is already inserted in the official recipe you have to consider it illegal due that minimal version, if you remove it, for who is using, Conan it will raise an ConanException, because it's no longer valid. Instead, you can raise an ConanInvalidConfiguration, which better in terms of filtering for CI, and it will not change the package ID, as the option is still there.

File: recipes/glib/all/conanfile.py
Comment (uilianries) [reply recipes/glib/all/conanfile.py:73]: If you apply the previous suggestion, you can revert this change.

Comment (prince-chrismc) [reply recipes/glib/all/conanfile.py]:
```suggestion
        if self.options.with_elf:
```

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
to retrigger ci... change the title or opening comment

**Review by SSE4 — APPROVED** [review]

**Review by SpaceIm — APPROVED** [review]

**Pull request merged**
